text/0031-integ-testing.md

```diff
@@ -0,0 +1,542 @@
+
+```ts
+new AwsAssertion(test, 'GetObject', {
+  request: AwsAssertionCall.S3.getObject({ ... }),
+  returns: { Data: ... },
+});
+
+new InvokeLambda(test, 'Invoke', {
+  lambda: functionArn,
+  throws: ...
+});
+```
+
+The `AwsAssertion` also allows multiple calls to be chained. The following calls Amazon S3 `GetObject` API
+based on the value returned from an Amazon DynamoDB `GetItem` call.
+
+```ts
+const response = new AwsAssertion(test, 'GetItem', {
+  request: AwsAssertionCall.DynamoDB.getItem({ ... }),
+});
+
+new AwsAssertion(test, 'GetObject', {
+  request: AwsAssertionCall.S3.getObject({ bucket: ..., key: response.Item.Payload }),
+});
+```
+
+When canned assertions are insufficient, `apptest` also provides a mechanism to write custom assertions.
+Custom assertions are simply AWS Lambda Functions, and in the CDK, these are any construct that implements `IFunction`.
+
+```ts
+import { Code, Function, Runtime } from '@aws-cdk/aws-lambda';
+
+// Checks that a standard header is present in the given S3 object
+class StdHeaderAssertion extends Function {
+  constructor(scope: Construct, id: string, props: StdHeaderAssertionProps) {
+    super(scope, id, {
+      code: Code.fromAsset('./standard-header-assertion.js'),
+      runtime: Runtime.NODEJS_12_X,
+      env: {
+        BUCKET_NAME: props.bucket.bucketName,
+        KEY: props.bucket.key
+      },
+    });
+  }
+}
+```
+
+The assertion passes or fails based on the success or failure of the Lambda's handler code.
+In both cases, it can also return an optional message, that will be made available in the test result.
+
+To invoke this assertion, you can simply use the `invokeLambda()` API -
+
+```ts
+const hdrAssertion = new StdHeaderAssertion(...);
+
+new InvokeLambda(test, {
+  lambda: hdrAssertion,
+});
+```
+
+> RFC Note: See [Appendix A](#appendix-a---assertions) for details on the underlying design.
+
+### Test snapshots
+
+On first run of a apptest test, a snapshot will be produced that *must be checked into source tree*.
+This is a snapshot of the [cloud assembly] containing both the CDK app (being tested) and the test
+assertions. The snapshot will be placed at a folder relative to the test file.
+
+```
+test
+├── repo.apptest.ts
+└── repo.apptest.snap
+    ├── manifest.json
+    ├── RepoStack.template.json
+    ├── ...
+    └── ...
+```
+
+On subsequent runs of this test, if the synthesized cloud assembly matches the
+snapshot, apptest will skip deployment and will not run any assertions. The test will be considered
+to have passed. This behaviour can be overridden either by setting the `ignoreSnapshot` option for
+individual tests or by setting the `CDK_APPTEST_IGNORE_SNAPSHOT` environment variable and will apply
+to all tests executed in that environment.
+
+However, when the synthesized cloud assembly does not match the snapshot, the test will be executed
+in full and the snapshot will be updated. As before, the changes to the snapshots must be checked into
```

Comment:
Executing the test can only happen in an environment that has access to an AWS account, which is normally not the case for CI builds. So I think we should distinguish between re-evaluation and just snapshot assertion.